crates/video/src/main.rs

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+mod stream;
+mod vhu_video;
+mod vhu_video_thread;
+mod video;
+mod video_backends;
+
+use std::path::Path;
+use std::sync::{Arc, RwLock};
+
+use clap::Parser;
+use log::{info, warn};
+use thiserror::Error as ThisError;
+use vhost::{vhost_user, vhost_user::Listener};
+use vhost_user_backend::VhostUserDaemon;
+use vhu_video::{BackendType, VuVideoBackend};
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+pub(crate) type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, ThisError)]
+pub(crate) enum Error {
+    #[error("Could not create backend: {0}")]
+    CouldNotCreateBackend(vhu_video::VuVideoError),
+    #[error("Could not create daemon: {0}")]
+    CouldNotCreateDaemon(vhost_user_backend::Error),
+    #[error("Failed creating listener: {0}")]
+    FailedCreatingListener(vhost_user::Error),
+}
+
+#[derive(Clone, Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct VideoArgs {
+    /// Unix socket to which a hypervisor connects to and sets up the control path with the device.
+    #[clap(short, long)]
+    socket_path: String,
+
+    /// Path to the video device file. Defaults to /dev/video0.
+    #[clap(short = 'd', long, default_value = "/dev/video0")]
+    v4l2_device: String,
+
+    /// Video backend to be used.
+    #[clap(short, long)]
+    #[clap(value_enum)]
+    backend: BackendType,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct VuVideoConfig {
+    pub socket_path: String,
+    pub v4l2_device: String,
+    pub backend: BackendType,
+}
+
+impl TryFrom<VideoArgs> for VuVideoConfig {
+    type Error = Error;
+
+    fn try_from(args: VideoArgs) -> Result<Self> {
+        // Divide available bandwidth by the number of threads in order
+        // to avoid overwhelming the HW.
+        let socket_path: String = args.socket_path.trim().to_string();
+        let v4l2_device: String = args.v4l2_device.trim().to_string();
+        let backend: BackendType = args.backend;
+
+        Ok(VuVideoConfig {
+            socket_path,
+            v4l2_device,
+            backend,
+        })
+    }
+}
+
+impl VuVideoConfig {
+    /// Return the path of the unix domain socket which is listening to
+    /// requests from the guest.
+    pub fn get_socket_path(&self) -> String {
+        String::from(&self.socket_path)
+    }
+}
+
+pub(crate) fn start_backend(config: VuVideoConfig) -> Result<()> {
+    loop {
```

Comment:
We should maybe find a standard way for all the vhost-device daemons regarding looping serving requests.

I think currently looping by restarting daemons is a bit hacky since the socket will actually killed and reinitialized, leaving a window where it becomes unavailable...

Reply:
At the moment, this loop is in line with what other backends do. If we decide to modify the strategy, I would be happy to prepare a separate PR with a proposal that applies to all backends to keep them consistent.

Reply:
Except for scsi (I just realized), that it has no loop... do you think it would be better to remove this loop for now?

Reply:
My comment was mostly reflecting on the current situation. This PR is fine with or without loop. We should just eventually figure out how to prevent each backend to re-decide on this :). Thats of course out-of-scope for this PR.